src/main/java/com/limechain/runtime/WasmSectionUtils.java

```diff
@@ -1,124 +1,120 @@
+        UnknownCustomSection runtimeApis = (UnknownCustomSection) moduleWithCustomSections
+            .customSection(RUNTIME_APIS_SECTION_NAME);
+        UnknownCustomSection runtimeVersion = (UnknownCustomSection) moduleWithCustomSections
+            .customSection(RUNTIME_VERSION_SECTION_NAME);
 
         // If we're missing any of the two custom sections, fallback to `Core_runtime`
         if (Objects.isNull(runtimeApis) || Objects.isNull(runtimeVersion)) {
             return null;
         }
 
         // Read as many api versions as there are (we don't know the length for some reason).
-        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.content());
+        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.bytes());
 
         // Construct the runtime version partially uninitialized,
         // and then immediately set its apis we've read from the custom section.
-        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.content()));
+        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.bytes()));
         version.setApis(apiVersions);
 
         return version;
     }
 
     /**
-     * Finds the "runtime_apis" custom section in the given wasm binary
+     * Parses the import section of a wasm blob and extracts the "memory" import if present.
+     *
+     * @param wasmBinary the wasm blob
+     * @return the parsed {@link org.wasmer.ImportObject.MemoryImport} if present,
+     * a default one with 24 initial pages otherwise
      */
-    @Nullable
-    public WasmCustomSection findRuntimeApisCustomSection(byte[] wasmBinary) {
-        return findCustomSection(wasmBinary, RUNTIME_APIS_SECTION_NAME);
+    public ImportObject.MemoryImport parseMemoryFromBinary(byte[] wasmBinary) {
+        Module moduleWithSections = toModuleWithSections(
+            wasmBinary, SectionId.IMPORT, SectionId.MEMORY, SectionId.EXPORT);
+        int initialPagesLimit = DEFAULT_MEMORY_PAGES;
+
+        // Per Runtime spec only one memory should be available for each module.
+        if (isMemorySectionValid(moduleWithSections.memorySection())
+            && isExportSectionValid(moduleWithSections.exportSection())) {
+            initialPagesLimit = moduleWithSections.memorySection().getMemory(0).memoryLimits().initialPages();
+        } else if (moduleWithSections.importSection() != null) {
+            Optional<Import> parsedMemoryImport = moduleWithSections.importSection().stream()
+                .filter(i -> i.name().equals(MEMORY_IMPORT_NAME))
+                .findFirst();
+            if (parsedMemoryImport.isPresent()) {
+                initialPagesLimit = ((MemoryImport) parsedMemoryImport.get()).limits().initialPages();
+            }
+        }
+
+        return new ImportObject.MemoryImport(ENV_MODULE_NAME, initialPagesLimit, false);
```

Comment:
I see we're able to obtain the `shared` attribute as well from the `MemoryLimits`. If it's not always false per spec, I'd also add that to our `ImportObject.MemoryImport`.

Even if it is always false per spec, it would be okay to check that in actuality.

Reply:
> Even if it is always false per spec, it would be good to check that in actuality.

I agree with you on this one.

Side note, in our host we're loading only a single instance of a WASM module at a time (old one is scrapped when new one needs to be instantiated). This means that a shared memory would not be useful/should not exist. The `shared` flag should be true when multiple modules use the same memory in order to avoid collisions.

Reply:
I see, that's clear. Soon enough though, this observation
> in our host we're loading only a single instance of a WASM module at a time (old one is scrapped when new one needs to be instantiated)

might not be true. For now it is, but we'll see how this evolves.